File: JetpacReloaded/C# version/JetpacReloaded/Fuel.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System.Collections.Generic;

namespace JetpacReloaded
{
    class Fuel : BaseObject
    {
        private bool landedOnLedge = false;

        public Fuel(Texture2D image)
        {
            m_image = image;
            m_width = image.Width;
            m_height = image.Height + 2;
            m_screenLocation = new Vector2(rand.Next(0, 750), -30);
        }

        public void Update(List<Ledge> ledges)
        {
            m_rect = new Rectangle((int)m_screenLocation.X, (int)m_screenLocation.Y, m_width, m_height);

            if (!landedOnLedge)
            {
                foreach (Ledge ledge in ledges)
                {
                    if (!m_rect.Intersects(ledge.LedgeRect))
                    {
                        m_screenLocation.Y += 0.2f;
                    }
                    else
                    {
                        landedOnLedge = true;
                    }
                }
            }
        }

        public void UpdatePosition(int x, int y)
        {
            m_screenLocation.X = x;
            m_screenLocation.Y = y;
        }

        public bool LowerFuel()
        {
            m_screenLocation.X = 440;
            if (m_screenLocation.Y < 450)
            {
                m_screenLocation.Y++;
            }
            return m_screenLocation.Y > 448;
        }

        public Rectangle FuelRect
        {
            get { return m_rect; }
        }
    }
}

File: JetpacReloaded/Typescript version/HTML5Canvas/ts/fuel.ts
import { Rectangle } from "../ts_lib/rectangle.js";
import { Texture2D } from "../ts_lib/spritebatch.js";
import { Vector2 } from "../ts_lib/vector2.js";
import { BaseObject } from "./baseobject.js";
import { Ledge } from "./ledge.js";

export class Fuel extends BaseObject {
    private landedOnLedge: boolean = false;
    constructor(image: Texture2D) {
        super();
        this.m_image = image;
        this.m_width = image.width;
        this.m_height = image.height + 2;
        this.m_screenLocation = new Vector2(this.rand.Next(0, 750), -30);
    }
    public Update(ledges: Array<Ledge>): void {
        this.m_rect = new Rectangle(this.m_screenLocation.X, this.m_screenLocation.Y, this.m_width, this.m_height);
        if (!this.landedOnLedge) {
            ledges.forEach(function (ledge) {
                if (!this.m_rect.Intersects(ledge.LedgeRect)) {
                    this.m_screenLocation.Y += 0.2;
                }
                else {
                    this.landedOnLedge = true;
                }
            });
        }
    }
    public UpdatePosition(x: number, y: number): void {
        this.m_screenLocation.X = x;
        this.m_screenLocation.Y = y;
    }
    public LowerFuel(): boolean {
        this.m_screenLocation.X = 440;
        if (this.m_screenLocation.Y < 450) {
            this.m_screenLocation.Y++;
        }
        return this.m_screenLocation.Y > 448;
    }
    public get FuelRect(): Rectangle {
        return this.m_rect;
    }
}
